File: Training #2/C++/Juanes/jollyjumpers.cpp
#include <cstdio>
#include <vector>
#include <cmath>

using namespace std;

bool jolly_jumper(const int n, const vector<int> & numbers) {

    if (n < 2) {
        return true;
    }

    vector<int> differences(n - 1, 0);

    for (int i = 0; i < n - 1; i++) {
        int dif = abs(numbers[i] - numbers[i + 1]);

        if (dif >= 1 && dif < n) {
            differences[dif - 1] = 1;
        }
    }

    for (int i = 0; i < n - 1; i++) {

        if (differences[i] == 0) {
            return false;
        }
    }

    return true;
}

int main() {
    int n;
    while (scanf("%d", &n) != EOF) {
        vector<int> numbers(n);
        for (int i = 0; i < n; i++) {
            scanf("%d", &numbers[i]);
        }

        if (jolly_jumper(n, numbers)) {
            printf("Jolly\n");
        } else {
            printf("Not jolly\n");
        }
    }
    return 0;
}

File: Trainings/Training #2/Python/Juanes/jollyjumpers.py
from sys import stdin

def jolly_jumper(n, numbers):
    if n < 2:
        return True


    differences = [0] * (n-1)

    for i in range(n-1):
        dif = abs(numbers[i] - numbers[i+1])
        if 1 <= dif < n:
            differences[dif - 1] = 1


    for i in range(n-1):
        if differences[i] == 0:
            return False

    return True

def main():
    input = stdin.read().strip()
    lines = input.splitlines()

    for line in lines:
        if not line:
            continue

        numbers = line.split()


        numbers = [int(num) for num in numbers]

        n = numbers[0]
        sequence = numbers[1:]

        if jolly_jumper(n, sequence):
            print("Jolly")
        else:
            print("Not jolly")

main()